oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/pipelined/ThreadMonitor.java

```diff
@@ -0,0 +1,134 @@
+         *
+         * @param threadMonitor The thread monitor where to register new threads.
+         */
+        public AutoRegisteringThreadFactory(ThreadMonitor threadMonitor) {
+            this.threadMonitor = threadMonitor;
+            this.delegate = Executors.defaultThreadFactory();
+        }
+
+        @Override
+        public Thread newThread(@NotNull Runnable r) {
+            Thread t = delegate.newThread(r);
+            threadMonitor.registerThread(t);
+            return t;
+        }
+    }
+
+    private static final Logger LOG = LoggerFactory.getLogger(ThreadMonitor.class);
+    private final CopyOnWriteArraySet<Thread> threadIds = new CopyOnWriteArraySet<>();
+
+    private final ThreadMXBean threadMXBean = ManagementFactory.getThreadMXBean();
+    private final List<GarbageCollectorMXBean> gcMxBeans = ManagementFactory.getGarbageCollectorMXBeans();
+    private final MemoryMXBean memoryMxBean = ManagementFactory.getMemoryMXBean();
+
+    private final Stopwatch stopwatch = Stopwatch.createUnstarted();
+
+    public void start() {
+        stopwatch.start();
+    }
+
+    public void registerThread(Thread thread) {
+        threadIds.add(thread);
+    }
+
+    public void unregisterThread(Thread thread) {
+        threadIds.remove(thread);
+    }
+
+    public void printStatistics() {
+        printStatistics("Thread/Memory report");
+    }
+
+    public void printStatistics(String heading) {
+        long timeSinceStartMillis = stopwatch.elapsed().toMillis();
+
+        StringBuilder sb = new StringBuilder(heading + ". Time since start of monitoring: " + stopwatch + "\n");
+
+        // Memory usage
+        sb.append(String.format("  Heap memory usage: %s, Non-heap memory usage: %s\n",
+                memoryMxBean.getHeapMemoryUsage(), memoryMxBean.getNonHeapMemoryUsage()));
+
+        // Garbage collection
+        for (GarbageCollectorMXBean gcBean : gcMxBeans) {
+            sb.append(String.format("  Collector: %s, collectionCount: %d, collectionTime: %d ms (%.2f%%)\n",
+                    gcBean.getName(), gcBean.getCollectionCount(), gcBean.getCollectionTime(),
+                    FormattingUtils.safeComputePercentage(gcBean.getCollectionTime(), timeSinceStartMillis))
+            );
+        }
+
+        // Thread CPU usage
+        threadIds.stream().sorted(Comparator.comparing(Thread::getId)).forEach(thread -> {
+            long threadCpuTimeMillis = threadMXBean.getThreadCpuTime(thread.getId()) / 1_000_000;
+            long threadUserTimeMillis = threadMXBean.getThreadUserTime(thread.getId()) / 1_000_000;
+            double threadCpuTimePercentage = FormattingUtils.safeComputePercentage(threadCpuTimeMillis, timeSinceStartMillis);
+            double threadUserTimePercentage = FormattingUtils.safeComputePercentage(threadUserTimeMillis, timeSinceStartMillis);
+            sb.append(String.format("  Thread %d/%s - cpuTime: %d (%.2f%%), userTime: %d (%.2f%%)\n",
```

Comment:
Thread names have a variable size... I usually put the (most) variable size entries at the end of the line. So it would be cpuTime, userTime, thread id / name

Reply:
I prefer to have the thread name at the start, reads more naturally. But I added some basic formatting to align the fields. Have a look at the updated table in the PR description, it is more readable now.